### ISSUE TITLE
LSP test framework improvements

### ISSUE THREAD

**Issue by christianparpart** [ethereum/solidity#12601]
After some discussions with @Marenz and now @chriseth, I'm writing down what we agreed on:

In order to avoid messing with file offsets, the idea is to reference cursor locations and source location ranges via Solidity comments.
Using one or more `^` below a line to mark the line above at these columns as a source location.

If it's a single `^`, then that location can be used as "position" (e.g. for client's text cursor), and a sequence of one or more `^` can be used to express a source location range.

Right next to the sequence of  `^`('s) will be the ID, starting with an `@` and followed by an identifier. 

Multiline source locations will have a single `^( @Identifier` marking the first line and a `^) @Identifier` marking the last line.

The new test framework should be similar to `isoltest`, having .sol files for each small test case. Multiple files are split by `// ==== ileA.sol` lines, if needed.

The format for the request(s) and expectation(s) can be similar to the following example:

```solidity
contract C {
    function f(uint256 var1) public pure returns { return 2 * var1; }
                   //  ^^^^ @var1-def                         ^ @use-of-var1
}
// ----
// send: textDocument/definition
// params: { @use-of-var1 }
// reply.result.0: { @var1-def }
```

On failure, an editor could be popped up (like isoltest) to interactively fix the mismatch, or maybe provide an auto-fix iff(!) the actual location is visualized.

**Comment by chriseth** [comment]
In addition to that, we should check if it is possible to make the expectations even more condensed:
```
// ----
// textDocument/definition @use-of-n
// -> @n-def
```
or something like that

**Comment by christianparpart** [comment]
this might work but not for all request/response cases. however, responses holding an array of locations could be expressed easily dense, too:
```
// -> (@n-def, @p-def, @q-def)
```

(or something like that)

**Comment by Marenz** [comment]
Looking at the test, I can see that we assert the diagnostics right after opening a file, I figured we should have that capability in this syntax too. 
This is what I currently came up with:

```
// SPDX-License-Identifier: UNLICENSED
pragma solidity >=0.8.0;

contract MyContract
{
    constructor()
    {
        uint unused; // [Warning 2072] Unused local variable.
   //   ^^^^^^^^^^^ @unusedVariable
    }
}

contract D
{
    function main() public payable returns (uint)
                                       //   ^^^^ @unusedReturnVariable
    {
        MyContract c = new MyContract();
   //   ^^^^^^^^^^^^ @unusedContractVariable
    }
}
// ----
// diagnostics:
//  - 6321 @unusedReturnVariable
//  - 2072 @unusedVariable
//  - 2072 @unusedContractVariable
// textDocument/didChange @unusedVariable+ ""
// diagnostics:
//  - 6321 @unusedReturnVariable
//  - 2072 @unusedContractVariable

```
the little `+` there means "the location of that variable `+1` because we want to replace the variable + one more (the `;`) with the string following. We have that situation also quiet a bit in the current cases. Could be extended to `+<number>` to be more generic, though I don't want to have too many raw numbers in there again.

The number in the diagnostic list is simply the reported error code.

**Comment by chriseth** [comment]
Is this feature relevant for anything except the didChange handler? Adding state and actions (that are not just queries) to the systems seem like a rather complex change that we maybe don't even need - and it's also difficult to understand what happens in the tests.

**Comment by Marenz** [comment]
Well, every time a change is done or a file opened we get the reported diagnostics through the LSP server and in every test every diagnostic is currently exactly checked & validated to be what we expect it to be, e.g. "unused variable" for a certain location/range. I saw it as part of the tests really.
The particular test shown here removes the unused variable and checks the new diagnostics after the change, expecting it to be gone. 
So it's like an before/after here. Though there could be more calls and state changes added of course.

**Comment by Marenz** [comment]
This has been resolved by #12724
